### DISCUSSION TITLE
BPMs Misalignemts

### DISCUSSION THREAD

**Discussion by elafmusa** [atcollab/at#799] [Python AT]
Dear AT group,

For simulation i am working on, I want to attach BPMs to quadrupoles and assign horizontal and vertical displacement errors to some quadrupoles. The BPM indices are the quadrupole indices + 1 (following beam direction).

I would like to cross-check with you whether the approach I followed to define the new reference orbit after assigning the errors is correct and whether I made any mistakes:

The approach:

1. Define field offset to the BPMs:
```
 for i in bpm_indices:
      ring[i].Offset = [0,0]

```
2-Assign hor. and vert. shift errors to some quadrupoles.

3-Assign quadrupole errors to the BPM offset field:

```

bpm_Offsetx = []
bpm_Offsety = []
for i in quads_indices:
      if hasattr(ring[i], 'T2'):
          bpm_Offsetx.append(ring[i].T2[0])
           bpm_Offsety.append(ring[i].T2[2])
      else:
            bpm_Offsetx.append(0)
            bpm_Offsety.append(0)


for i in range(len(bpm_indices)):
      ring[bpm_indices[i]].Offset[0] = bpm_Offsetx[i]
      ring[bpm_indices[i]].Offset[1] = bpm_Offsety[i]

```
4-Obtain the new horizontal and vertical reference orbits:

```
hor_ref_orbit = [ring[i].Offset[0] for i in bpm_ind]
ver_ref_orbit = [ring[i].Offset[1] for i in bpm_ind]
```

Thank you in advance,

**Comment by oscarxblanco** [comment]
Dear @elafmusa , it seems correct as long as the number of `bpm_indices`and `quads_indices` is the same and they are ordered as you need, so there is a bijection (a one  to one correspondence).

For other cases it looks like it will run into errors.

**Comment by elafmusa** [comment]
Dear @oscarxblanco,

Thank you for your response!

Could you please clarify the differences between the approach I used and the one provided in the example below?

Instead of:

hor_ref_orbit = [ring[i].Offset[0] for i in bpm_ind]
ver_ref_orbit = [ring[i].Offset[1] for i in bpm_ind]

The alternative approach is:

[elemdata0, beamdata, elemdata] = at.get_optics(ring0, quadrupoles_ind)
hor_ref_orbit = elemdata.closed_orbit[:, 0]
ver_ref_orbit = elemdata.closed_orbit[:, 2]

Thank you for your help!

**Comment by oscarxblanco** [comment]
Dear @elafmusa , they are different.

The first method is giving you two lists with the values of the offsets on the **BPMs**. Typically the BPM length is zero, so entry and exit faces are the same.

The second one calculates the closed orbit and gives you the transverse (hor. and ver.) position of **beam** at the entry face of the quads. The closed orbit is calculated over `ring0` which depending on the type of calculation (4D,synch,6D) will give you a different result. On top of that your ring could have magnetic errors and misalignments that will change the closed orbit, and therefore the transverse positions on your quads.

Whether you want to call `reference_orbit` one or the other depends on what you would like to do after that.
